### PR TITLE
Introduce interconnections package

### PR DESCRIPTION
**Pull request by kayrus**
For the https://github.com/sapcc/networking-interconnection service

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
I would like to postpone this until #97 is merged and then add the context arguments straight away to this PR.

In the meantime please fix the golangci-lint lints.

**Comment by kayrus** [comment]
@SuperSandro2000 any ideas on how to fix this?

```
  Error: networking/v2/interconnections/requests.go:60:25: response body must be closed (bodyclose)
  Error: networking/v2/interconnections/requests.go:94:26: response body must be closed (bodyclose)
  Error: networking/v2/interconnections/requests.go:101:28: response body must be closed (bodyclose)
  Error: networking/v2/interconnections/requests.go:130:25: response body must be closed (bodyclose)
```
this is a common pattern and the body is closed during the `Extract()` method call.

**Comment by SuperSandro2000** [comment]
> @SuperSandro2000 any ideas on how to fix this?

Those are false positives, I just ignored them [`dcab157` (#97)](https://github.com/sapcc/gophercloud-sapcc/pull/97/commits/dcab1579a2948487eb9c17d8c602ace8716a16d8)

> this is a common pattern and the body is closed during the Extract() method call.

Yep and the linter can't understand that rather complex interaction.

**Comment by kayrus** [comment]
@SuperSandro2000 currently this PR blocks further tasks. Can we merge it right away? I can adjust the #97 accordingly once merged.

**Review by velp — APPROVED** [review]
LGTM

File: networking/v2/bgpvpn/interconnections/results.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+// Extract will extract a response object from a result.
+func (r commonResult) Extract() (*Interconnection, error) {
+	var s Interconnection
+	err := r.ExtractInto(&s)
+	return &s, err
+}
+
+func (r commonResult) ExtractInto(v interface{}) error {
```
Comment (SuperSandro2000) [inline networking/v2/bgpvpn/interconnections/results.go:41]:
```suggestion
func (r commonResult) ExtractInto(v any) error {
```

File: networking/v2/bgpvpn/interconnections/results.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+// InterconnectionPage struct, and extracts the elements into a slice of
+// Interconnection structs. In other words, a generic collection is mapped into
+// a relevant slice.
+func ExtractInterconnections(r pagination.Page) ([]Interconnection, error) {
+	var s []Interconnection
+	err := ExtractInterconnectionssInto(r, &s)
+	return s, err
+}
+
+func ExtractInterconnectionssInto(r pagination.Page, v interface{}) error {
```
Comment (SuperSandro2000) [inline networking/v2/bgpvpn/interconnections/results.go:154]:
```suggestion
func ExtractInterconnectionssInto(r pagination.Page, v any) error {
```

File: networking/v2/bgpvpn/interconnections/requests.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	//nolint:bodyclose // already handled by gophercloud
+	resp, err := c.Get(getURL(c, id), &r.Body, nil)
+	_, r.Header, r.Err = gophercloud.ParseResponse(resp, err)
+	return
+}
+
+// CreateOptsBuilder allows extensions to add additional parameters to the
+// Create request.
+type CreateOptsBuilder interface {
+	ToInterconnectionCreateMap() (map[string]interface{}, error)
```
Comment (SuperSandro2000) [inline networking/v2/bgpvpn/interconnections/requests.go:83]:
```suggestion
	ToInterconnectionCreateMap() (map[string]any, error)
```

File: networking/v2/bgpvpn/interconnections/requests.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	ProjectID               string `json:"project_id,omitempty"`
+	Type                    string `json:"type,omitempty"`
+	LocalResourceID         string `json:"local_resource_id"`
+	RemoteResourceID        string `json:"remote_resource_id"`
+	RemoteRegion            string `json:"remote_region"`
+	RemoteInterconnectionID string `json:"remote_interconnection_id,omitempty"`
+}
+
+// ToInterconnectionCreateMap formats a CreateOpts into a map.
+func (opts CreateOpts) ToInterconnectionCreateMap() (map[string]interface{}, error) {
```
Comment (SuperSandro2000) [inline networking/v2/bgpvpn/interconnections/requests.go:98]:
```suggestion
func (opts CreateOpts) ToInterconnectionCreateMap() (map[string]any, error) {
```

File: networking/v2/bgpvpn/interconnections/requests.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	//nolint:bodyclose // already handled by gophercloud
+	resp, err := c.Delete(deleteURL(c, id), nil)
+	_, r.Header, r.Err = gophercloud.ParseResponse(resp, err)
+	return
+}
+
+// UpdateOptsBuilder allows extensions to add additional parameters to the
+// Update request.
+type UpdateOptsBuilder interface {
+	ToInterconnectionUpdateMap() (map[string]interface{}, error)
```
Comment (SuperSandro2000) [inline networking/v2/bgpvpn/interconnections/requests.go:126]:
```suggestion
	ToInterconnectionUpdateMap() (map[string]any, error)
```

File: networking/v2/bgpvpn/interconnections/requests.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+// UpdateOpts represents options used to update an interconnection.
+type UpdateOpts struct {
+	Name  *string `json:"name,omitempty"`
+	State *string `json:"state,omitempty"`
+}
+
+// ToInterconnectionUpdateMap formats an UpdateOpts into a map.
+func (opts UpdateOpts) ToInterconnectionUpdateMap() (map[string]interface{}, error) {
```
Comment (SuperSandro2000) [inline networking/v2/bgpvpn/interconnections/requests.go:136]:
```suggestion
func (opts UpdateOpts) ToInterconnectionUpdateMap() (map[string]any, error) {
```

File: networking/v2/bgpvpn/interconnections/requests.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+// Update allows interconnections to be updated.
+func Update(c *gophercloud.ServiceClient, id string, opts UpdateOptsBuilder) (r UpdateResult) {
+	b, err := opts.ToInterconnectionUpdateMap()
+	if err != nil {
+		r.Err = err
+		return
+	}
+	//nolint:bodyclose // already handled by gophercloud
+	resp, err := c.Put(updateURL(c, id), b, &r.Body, &gophercloud.RequestOpts{
+		OkCodes: []int{200},
```
Comment (SuperSandro2000) [inline networking/v2/bgpvpn/interconnections/requests.go:149]:
```suggestion
		OkCodes: []int{http.StatusOK},
```

File: networking/v2/bgpvpn/interconnections/requests.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interconnections
+
+import (
```
Comment (SuperSandro2000) [inline networking/v2/bgpvpn/interconnections/requests.go:17]:
```suggestion
import (
	"http"

```

**Comment by SuperSandro2000** [comment]
I've also tagged a release to make the gophercloud 2.0 transition easier. https://github.com/sapcc/gophercloud-sapcc/releases/tag/v1.0.0

**Pull request merged**